### DISCUSSION TITLE
Overwrite Conformers

### DISCUSSION THREAD

**Discussion by RMeli** [rdkit/rdkit#5188] [Q&A] [answered]
I have a function that takes a molecule with a conformer (`0`), randomly rotates the conformer, and store it as a new conformer (`1`) on the same molecule. However I noticed that when calling said function repeatedly on the same molecule, I was getting the same results despite the random rotation. It seems that RDKit does not overwrite conformers, but it does not warn either if I try to assign a new conformer with an ID that is already taken.

An MWE illustrating the issue (?) is the following:
```python
from rdkit.Chem import Chem, AllChem

mol = Chem.MolFromSmiles("COO")
molh = Chem.AddHs(mol)

ids = AllChem.EmbedMultipleConfs(molh, numConfs=2)
list(ids)

print("molh | conf 0\n", molh.GetConformer(0).GetPositions()[0])
print("molh | conf 1\n", molh.GetConformer(1).GetPositions()[0])

cmolh = Chem.Mol(molh, True) # Copy original molecule without conformers

# Add conf 0 of molh as conf 0 of cmolh
conf = molh.GetConformer(0)
conf.SetId(0)
cmolh.AddConformer(conf, assignId=False)
print("cmolh | conf 0\n", cmolh.GetConformer(0).GetPositions()[0])

# Try to overwrite conf 0 of cmolh with conf 1 of molh
conf = molh.GetConformer(1)
conf.SetId(0)  # Same ID; trying to overwrite...
cmolh.AddConformer(conf, assignId=False)
print("cmolh | conf 0\n", cmolh.GetConformer(0).GetPositions()[0])
```

The output of the above example is the following:
```
molh | conf 0
 [-0.74067015  0.0880062  -0.04217678]
molh | conf 1
 [-0.7319001   0.14769868  0.58309664]
cmolh | conf 0
 [-0.74067015  0.0880062  -0.04217678]
cmolh | conf 0
 [-0.74067015  0.0880062  -0.04217678]
```
As you can see, the output `cmolh | conf 0` is the same. However, I would expect it to be different since the first time I have `conf = molh.GetConformer(0)` and the second time I have `conf = molh.GetConformer(1)`.

The solution to this problem is to call `cmolh.RemoveConformer(0)` before assigning the new conformer:

```python
from rdkit.Chem import Chem, AllChem

mol = Chem.MolFromSmiles("COO")
molh = Chem.AddHs(mol)

ids = AllChem.EmbedMultipleConfs(molh, numConfs=2)
list(ids)

print("molh | conf 0\n", molh.GetConformer(0).GetPositions()[0])
print("molh | conf 1\n", molh.GetConformer(1).GetPositions()[0])

cmolh = Chem.Mol(molh, True) # Copy original molecule without conformers

# Add conf 0 of molh as conf 0 of cmolh
conf = molh.GetConformer(0)
conf.SetId(0)
cmolh.AddConformer(conf, assignId=False)
print("cmolh | conf 0\n", cmolh.GetConformer(0).GetPositions()[0])

# Try to overwrite conf 0 of cmolh with conf 1 of molh
conf = molh.GetConformer(1)
conf.SetId(0)
cmolh.RemoveConformer(0) # Explicitly remove previous conformer!
cmolh.AddConformer(conf, assignId=False)
print("cmolh | conf 0\n", cmolh.GetConformer(0).GetPositions()[0])
```

However, since it took me a while to figure out what was happening, I was wondering if this is the expected behaviour. Shouldn't RDKit complain in a way or another if we try to assign a new conformer to an existing ID instead of silently ignoring the update? Am I missing something obvious here?

**Comment by greglandrum** [comment, marked as answer]
I agree that the behavior is somewhat surprising, but this is the expected behavior: unless you tell `AddConformer()` to set the ID of the conformer, it will just use whatever you provide. It's up to you to make sure that the conformer IDs are unique.

Remember that you aren't "assigning a new conformer to an existing ID", you are *adding* a conformer.

The correct pattern if you want to replace a conformer with a particular ID is to remove the current conformer with that ID and then add the new one. _Note_: technically it doesn't matter which order you do these steps in, since it will remove the first conformer it finds with the particular ID, but it would definitely make for clearer code if you didn't rely on this behavior.

**Reply by RMeli** [reply]
Thank you for the clarifications @greglandrum !
